kimchi/src/index.rs

```diff
@@ -163,36 +172,35 @@ pub fn constraints_expr<F: FftField + SquareRootField>(
     domain: D<F>,
     chacha: bool,
     lookup_constraint_system: &Option<LookupConstraintSystem<F>>,
-) -> (Expr<ConstantExpr<F>>, alphas::Builder) {
+) -> (Expr<ConstantExpr<F>>, alphas::Alphas<F>) {
     // register powers of alpha so that we don't reuse them across mutually inclusive constraints
-    let mut powers_of_alpha = alphas::Builder::default();
+    let mut powers_of_alpha = alphas::Alphas::<F>::default();
 
     // gates
-    let largest_constraint_num = varbasemul::CONSTRAINTS;
-    let alphas = powers_of_alpha.register(ConstraintType::Gate, largest_constraint_num);
+    let highest_constraints = VarbaseMul::<F>::CONSTRAINTS;
+    powers_of_alpha.register(ConstraintType::Gate, highest_constraints);
 
-    let mut expr = poseidon::constraint(alphas.clone().take(poseidon::CONSTRAINTS));
-    expr += varbasemul::constraint(alphas.clone().take(varbasemul::CONSTRAINTS));
-    expr += complete_add::constraint(alphas.clone().take(complete_add::CONSTRAINTS));
-    expr += endosclmul::constraint(alphas.clone().take(endosclmul::CONSTRAINTS));
-    expr += endomul_scalar::constraint(alphas.clone().take(endomul_scalar::CONSTRAINTS));
+    let mut expr = Poseidon::default().constraint(&powers_of_alpha);
```

Comment:
It might be cleaner to have a function that returns an `impl trait Argument` than do the default dance on a dummy object. Wdyt?

Reply:
yeah in general I agree, but I think you also end up making things a bit too magical without really reducing the complexity. There's a trade off.